harness/determined/common/storage/shared.py

```diff
@@ -3,134 +3,97 @@
-                    # Special check for directory junctions, which appear as
-                    # symlinks but we want to recurse.
-                    lstat = srcentry.stat(follow_symlinks=False)
-                    if lstat.st_reparse_tag == stat.IO_REPARSE_TAG_MOUNT_POINT:
-                        is_symlink = False
-                if is_symlink:
-                    linkto = os.readlink(srcname)
-                    if symlinks:
-                        # We can't just leave it to `copy_function` because legacy
-                        # code with a custom `copy_function` may rely on copytree
-                        # doing the right thing.
-                        os.symlink(linkto, dstname)
-                        shutil.copystat(srcobj, dstname, follow_symlinks=not symlinks)
-                    else:
-                        # ignore dangling symlink if the flag is on
-                        if not os.path.exists(linkto) and ignore_dangling_symlinks:
-                            continue
-                        # otherwise let the copy occur. copy2 will raise an error
-                        if srcentry.is_dir():
-                            copytree(
-                                srcobj,
-                                dstname,
-                                symlinks,
-                                ignore,
-                                copy_function,
-                                dirs_exist_ok=dirs_exist_ok,
-                            )
-                        else:
-                            copy_function(srcobj, dstname)
-                elif srcentry.is_dir():
-                    copytree(
-                        srcobj,
-                        dstname,
-                        symlinks,
-                        ignore,
-                        copy_function,
-                        dirs_exist_ok=dirs_exist_ok,
-                    )
-                else:
-                    # Will raise a SpecialFileError for unsupported file types
-                    copy_function(srcobj, dstname)
-            # catch the Error from the recursive copytree so that we can
-            # continue with other files
-            except shutil.Error as err:
-                errors.extend(err.args[0])
-            except OSError as why:
-                errors.append((srcname, dstname, str(why)))
+# Based on shutil.copytree and shutil._copytree (for Python 3.8). Compared to the original
+# implementation this code delays creating new directories during traversal, such that dir
+# is created only when (1) selector(dir)==True or (2) selector(dir/subdir/path)==True.
+# The code is simplified to rely on default values for:
+# symlinks=False,
+# ignore=None,
+# copy_function=shutil.copy2,
+# ignore_dangling_symlinks=False.
+
+
+def _copytree(
+    entries: List,
+    src: str,
+    dst: str,
+    selector: Optional[Callable[[str], bool]],
+    src_root: str,
+    dirs_exist_ok: bool = False,
```

Comment:
nit: I'd remove this flag, since we literally always set it to not-the-default.

My reasoning is that "unused tooling can confuse the intention of a piece of code".